Fix #3205: [Rtl] High-fi Home Screen All Topics List

veena14cs opened the pull request:
## Explanation
This PR fixes #3205. 

## Screenshot LTR and RTL

<img width="144" alt="screenshot" src="https://user-images.githubusercontent.com/16301028/122271569-2a7fd300-cefd-11eb-8733-89198f6dd44f.png">  .... <img width="144" alt="screenshot" src="https://user-images.githubusercontent.com/16301028/122271627-3f5c6680-cefd-11eb-95c9-7e291c71e9e5.png">

<img width="256" alt="screenshot" src="https://user-images.githubusercontent.com/16301028/122271508-19cf5d00-cefd-11eb-9a8e-41967a5ef2b1.png">  ...  <img width="256" alt="screenshot" src="https://user-images.githubusercontent.com/16301028/122271710-54d19080-cefd-11eb-9815-2fe8ccbaa528.png">
`1324e566tr6


Tablet Portrait: 
<img width="160" alt="screenshot" src="https://user-images.githubusercontent.com/16301028/122339806-2389ac80-cf5f-11eb-94af-4d6a690a81de.png"> ....  <img width="160" alt="screenshot" src="https://user-images.githubusercontent.com/16301028/122339881-38fed680-cf5f-11eb-878a-bb5a8059b47f.png">

Landscape:

<img width="256" alt="screenshot" src="https://user-images.githubusercontent.com/16301028/124327096-30381280-dba5-11eb-83e8-1ad1d642f422.png">........<img width="256" alt="screenshot" src="https://user-images.githubusercontent.com/16301028/124327147-40e88880-dba5-11eb-9a40-d5b5cea4cb26.png">

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

veena14cs committed:
fixed rtl ui

veena14cs committed:
Fixed rtl ui

veena14cs committed:
Merge branch 'develop' into rtl-all-topics

veena14cs commented:
@rt4914 Can you please look into this PR. There is issue with Tablet landscape mode. I am not getting equal spacing around the items. I tried setting margins in different ways but its not working.

veena14cs commented:
<img width="256" alt="screenshot" src="https://user-images.githubusercontent.com/16301028/122340471-ed006180-cf5f-11eb-8f14-b60142b6efa9.png">

rt4914 commented:
@veena14cs I haven't been able to solve this today remained stuck on it. Will try again tomorrow with fresh mind.

veena14cs commented:
> @veena14cs I haven't been able to solve this today remained stuck on it. Will try again tomorrow with fresh mind.

Ok @rt4914 did you find any fix?

veena14cs committed:
Update selection_interaction_item.xml

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Update selection_interaction_item.xml

rt4914 commented:
> > @veena14cs I haven't been able to solve this today remained stuck on it. Will try again tomorrow with fresh mind.
> 
> Ok @rt4914 did you find any fix?

@veena14cs Haven't been able to find any fix. Also I am unable to spend a lot of time on this so you might need to try some different methods to find the problem and then solve it.

veena14cs commented:
> > > @veena14cs I haven't been able to solve this today remained stuck on it. Will try again tomorrow with fresh mind.
> > 
> > 
> > Ok @rt4914 did you find any fix?
> 
> @veena14cs Haven't been able to find any fix. Also I am unable to spend a lot of time on this so you might need to try some different methods to find the problem and then solve it.

Thanks @rt4914 I will try out different solutions. 
@BenHenning do you have any suggestions on this.

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Update return_to_topic_button_item.xml

veena14cs committed:
Update return_to_topic_button_item.xml

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

BenHenning commented:
@veena14cs I'm probably not going to have bandwidth to look into this this week. You may need to step debug into the actual UI components that we're using to better understand the spacing logic and why the observed behavior is happening in order to figure out how to fix it (this is what I would do if I were to dig into this problem in detail). It's difficult for us to help without more context, so this seems to be the next step to take.

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Update GridSpacingItemDecoration.kt

veena14cs committed:
fixed landscape view

veena14cs commented:
> @veena14cs I'm probably not going to have bandwidth to look into this this week. You may need to step debug into the actual UI components that we're using to better understand the spacing logic and why the observed behavior is happening in order to figure out how to fix it (this is what I would do if I were to dig into this problem in detail). It's difficult for us to help without more context, so this seems to be the next step to take.

Thanks Ben! I have fixed the issue, it will be good if you can please review this PR. I have added tablet landscape mode screenshot in PR description for reference. The issue was there is no equal spacing for the items. Solution was to add equal spacing for all the items.

@rt4914 PTAL

veena14cs committed:
Update GridSpacingItemDecoration.kt

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
resolved conflicts

rt4914 commented on app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt:
```diff
@@ -0,0 +1,131 @@
+import org.oppia.android.R
+import org.oppia.android.app.home.HomeFragmentPresenter
+
+class GridSpacingItemDecoration(context: Context) : ItemDecoration() {
```
Add Kdoc for this file.

rt4914 replied:
Also please add reference from where this code has been used.

veena14cs replied:
Done.

rt4914 replied:
Nit
```
/** 
 * An [RecyclerView.ItemDecoration] that adds spacing between cells in a [GridLayoutManager].
 * 
 * Reference: https://stackoverflow.com/a/30701422/4500099
 */
```

rt4914 replied:
@veena14cs PTAL

veena14cs replied:
Done.

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

rt4914 commented on app/src/main/java/org/oppia/android/app/home/HomeFragmentPresenter.kt:
```diff
@@ -77,6 +78,8 @@ class HomeFragmentPresenter @Inject constructor(
         else spanCount
       }
     }
+    binding.homeRecyclerView.addItemDecoration(GridSpacingItemDecoration(activity));
```
Remove `;` at the end.

veena14cs replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt:
```diff
@@ -0,0 +1,131 @@
+
+    if (viewType == HomeFragmentPresenter.ViewType.TOPIC_LIST.ordinal) {
+//
+      /* INVALID SPAN */if (spanCount < 1) return
```
Remove this comment and above `//` too.

veena14cs replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt:
```diff
@@ -0,0 +1,131 @@
+
+      if (spanCount == 4) {
+        outRect.left =
+          (spacing ) - spanIndex * (spacing ) / spanCount // spacing - column * ((1f / spanCount) * spacing)
```
Nit:
`(spacing) - spanIndex * (spacing) / spanCount // spacing - column * ((1f / spanCount) * spacing)`

veena14cs replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt:
```diff
@@ -0,0 +1,131 @@
+        outRect.right = tmp
+      }
+
+      if (spanIndex < spanCount) { // top edge
```
What does `top edge` means here?

veena14cs replied:
Top margin.

rt4914 commented on app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt:
```diff
@@ -0,0 +1,131 @@
+  }
+
+  private fun getItemSpanSize(parent: RecyclerView, childIndex: Int): Int {
+    val mgr = parent.layoutManager
```
What is `mgr`? Can we give a better name for clear understanding? to me it looks like `layoutManager`

veena14cs replied:
Done.

rt4914 reviewed (changes requested):
@veena14cs Suggested changes.

veena14cs committed:
updated changes

veena14cs committed:
Merge branch 'develop' into rtl-all-topics

veena14cs commented:
Thanks @rt4914 fixed changes PTAL.

rt4914 reviewed (changes requested):
@BenHenning Any idea how to approach towards adding test for this new `GridSpacingItemDecoration` file. I think its more of `Screenshot` testing part but still if you have any other suggestions.

BenHenning commented on app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt:
```diff
@@ -0,0 +1,135 @@
+    private const val VERTICAL = OrientationHelper.VERTICAL
+
+    @SuppressLint("NewApi", "WrongConstant")
+    private fun isLayoutRTL(parent: RecyclerView): Boolean {
```
Nit: ``isRtlLayout`` (switched the word order, and we should say 'Rtl' not 'RTL' per camel-casing guidelines in the style guide).

veena14cs replied:
Thanks! done.

BenHenning reviewed (commented):
@rt4914 @veena14cs could we maybe test this with Robolectric? If we lay out a test FrameLayout that contains the RecyclerView with the decorator attached, we should be able to create different situations to test the decoration.

I agree that Espresso + screenshot tests would probably be easier, but likely not thorough. Screenshot tests should verify visual correctness, not functional correctness. We should still try to use unit-level tests for that.

veena14cs committed:
Update GridSpacingItemDecoration.kt

rt4914 commented:
> @rt4914 @veena14cs could we maybe test this with Robolectric? If we lay out a test FrameLayout that contains the RecyclerView with the decorator attached, we should be able to create different situations to test the decoration.
> 
> I agree that Espresso + screenshot tests would probably be easier, but likely not thorough. Screenshot tests should verify visual correctness, not functional correctness. We should still try to use unit-level tests for that.

@veena14cs One of the ways we can test this is by using an item and this decorator such that if there is (decorator + item) then we can display 3 items in grid but if its only item without decorator then the grid shows 4 or more items which automatically proves that the decorator is taking taking up some space. We can use this for different qualifiers and check.

@veena14cs WDYT?

veena14cs commented:
> > @rt4914 @veena14cs could we maybe test this with Robolectric? If we lay out a test FrameLayout that contains the RecyclerView with the decorator attached, we should be able to create different situations to test the decoration.
> > I agree that Espresso + screenshot tests would probably be easier, but likely not thorough. Screenshot tests should verify visual correctness, not functional correctness. We should still try to use unit-level tests for that.
> 
> @veena14cs One of the ways we can test this is by using an item and this decorator such that if there is (decorator + item) then we can display 3 items in grid but if its only item without decorator then the grid shows 4 or more items which automatically proves that the decorator is taking taking up some space. We can use this for different qualifiers and check.
> 
> @veena14cs WDYT?

Just for clarification, should I test span count over here? There is already Roboelectric test `HomeSpanTest` and it passes all the checks.

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into rtl-all-topics

veena14cs commented:
This issue is getting fixed in #3452 . Once that is verified and merged will be closing this PR.

veena14cs commented:
This issue is fixed in #3452. Therefore, closing this PR.
